### ISSUE TITLE
Redis - Command timeout on MoveDelayed

### ISSUE THREAD

**Issue by PeterOscarsson** [blehnen/DotNetWorkQueue#42]
Hi 
I have recently got a lot of problems with timeouts.
This is for all of my applications where I use Redis Workqueues
There are about 10 of this at the same second in my log.

>>
StackExchange.Redis.RedisTimeoutException: Timeout performing EVAL (5000ms), next: EVAL, inst: 2, qu: 0, qs: 1, aw: False, bw: SpinningDown, rs: ReadAsync, ws: Idle, in: 4, in-pipe: 0, out-pipe: 0, last-in: 314, cur-in: 0, sync-ops: 2864, async-ops: 1, serverEndpoint: redis-qa.ad.impleoit.se:6379, conn-sec: 3087.67, aoc: 0, mc: 1/1/0, mgr: 10 of 10 available, clientName: adm-server-5488bf97c5-tzsz2(SE.Redis-v2.6.122.38350), IOCP: (Busy=0,Free=1000,Min=1,Max=1000), WORKER: (Busy=13,Free=32754,Min=4,Max=32767), POOL: (Threads=14,QueuedItems=51,CompletedItems=150671,Timers=57), v: 2.6.122.38350 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server, T defaultValue) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 2099
   at StackExchange.Redis.RedisDatabase.ScriptEvaluate(String script, RedisKey[] keys, RedisValue[] values, CommandFlags flags) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 1527
   at StackExchange.Redis.RedisDatabase.ScriptEvaluate(LoadedLuaScript script, Object parameters, CommandFlags flags) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 1542
   at DotNetWorkQueue.Transport.Redis.Basic.Lua.BaseLua.TryExecute(Object parameters)
   at DotNetWorkQueue.Transport.Redis.Basic.Lua.MoveDelayedToPendingLua.Execute(Int64 unixTime, Int32 count)
   at DotNetWorkQueue.Transport.Redis.Basic.CommandHandler.MoveDelayedRecordsCommandHandler.Handle(MoveDelayedRecordsCommand command)
   at DotNetWorkQueue.Transport.Redis.Basic.DelayedProcessingAction.Run(CancellationToken token)
at DotNetWorkQueue.Transport.Redis.Basic.Metrics.Decorator.DelayedProcessingActionDecorator.Run(CancellationToken token)
   at DotNetWorkQueue.Transport.Redis.Basic.Logging.Decorator.DelayedProcessingActionDecorator.Run(CancellationToken token)
   at DotNetWorkQueue.Queue.BaseMonitor.RunMonitor()
>>

**Comment by blehnen** [comment]
Hi,

There are a few options here.  Are you using multiple consumers?  Each one is probably looking for messages to move from the delay queue to the main queue.

1. You could increase the timeout from the default of 5000 ms on the redis connection string.  When testing, I tend to use 10000 or 15000, but that's because my redis server is running on slow/old hardware.
2. You can change the batch size from the default of 50 to something smaller, such as 10.

` 
var queueOptions = new RedisQueueTransportOptions(new SntpTimeConfiguration(),
    new DelayedProcessingConfiguration())
    {
        MoveDelayedMessagesBatchLimit = 10
     };

//inject into queue

using (var queueContainer = new QueueContainer<RedisQueueInit>(serviceRegister =>
                           AddInjectors(serviceRegister, queueOptions)))
            {

private static void AddInjectors(IContainer container, RedisQueueTransportOptions options)
{
    container.RegisterNonScopedSingleton(options);
}
`
3. You could change the monitor time from 1 second to something else, though that would increase the lag when processing delayed messages.  This is on the 'DelayedProcessingConfiguration' above.
4. If you are using multiple consumers, you could have 1 consumer check every 1 seconds (default), and the rest have a much longer check.  However, this is risky, as if the quick consumer goes down, your messages will be processed with an extra delay.

**Comment by PeterOscarsson** [comment]
Ok. The problem was due to the fact that I had moved the services to a server with less CPU count, so I had to tweek the thredpool a bit, due to throtling of new thread creation.

newMinWorker = 100;   // Considered safe.   Is number of CPUs as default (4, in my case)
newMinIOCP = 100;  // Considered safe.   Is number of CPUs as default (4, in my case)

System.Threading.ThreadPool.SetMinThreads(newMinWorker, newMinIOCP);

No more problems after this fix. :-)
